### PR TITLE
support amp training for detection models

### PR DESCRIPTION
**Pull request by xiaohu2015**
The pr is about https://github.com/pytorch/vision/issues/4509.
Since amp is supported on classification training, I also modify some files to support amp training on detetction models.

### CODE REVIEW

File: references/detection/engine.py [hunk 1]
```diff
@@ -44,8 +43,13 @@ def train_one_epoch(model, optimizer, data_loader, device, epoch, print_freq):
             sys.exit(1)
 
         optimizer.zero_grad()
-        losses.backward()
-        optimizer.step()
+        if scaler:
```
Comment (datumbox) [inline references/detection/engine.py:46]: Nit: Perhaps be explicit here as above:
```suggestion
        if scaler is not None:
```

File: references/detection/train.py [hunk 2]
```diff
@@ -245,6 +252,8 @@ def main(args):
                 "args": args,
                 "epoch": epoch,
             }
+            if args.amp:
+                checkpoint['scaler'] = scaler.state_dict()
```
Comment (datumbox) [inline references/detection/train.py:256]: Lint:
```suggestion
                checkpoint["scaler"] = scaler.state_dict()
```

**Review by datumbox — COMMENTED** [review]
There are some related [linter issues](https://app.circleci.com/pipelines/github/pytorch/vision/12547/workflows/60642811-7815-4c11-b125-fcec895a4b1b/jobs/992650), but overall it looks good.

@prabhat00155 could you have also a look since you did the classification one?

**Comment by xiaohu2015** [comment]
@datumbox Thanks. I have test the amp code, it works well.

**Comment by datumbox** [comment]
@xiaohu2015 There are a couple of more linter issues (spaces). Have a look a the [CI job](https://app.circleci.com/pipelines/github/pytorch/vision/12557/workflows/5d562542-9fc7-4087-87e1-74cf78ce2c90/jobs/993593), at the end it will show you the errors. For your convenience here are the things you need to change to keep it happy:
```
diff --git a/references/detection/train.py b/references/detection/train.py
index 5c50dcfa..ae13a32b 100644
--- a/references/detection/train.py
+++ b/references/detection/train.py
@@ -143,7 +143,7 @@ def get_args_parser(add_help=True):
 
     # Prototype models only
     parser.add_argument("--weights", default=None, type=str, help="the weights enum name to load")
-    
+
     # Mixed precision training parameters
     parser.add_argument("--amp", action="store_true", help="Use torch.cuda.amp for mixed precision training")
 
@@ -211,9 +211,9 @@ def main(args):
 
     params = [p for p in model.parameters() if p.requires_grad]
     optimizer = torch.optim.SGD(params, lr=args.lr, momentum=args.momentum, weight_decay=args.weight_decay)
-    
+
     scaler = torch.cuda.amp.GradScaler() if args.amp else None
-    
+
     args.lr_scheduler = args.lr_scheduler.lower()
     if args.lr_scheduler == "multisteplr":
         lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=args.lr_steps, gamma=args.lr_gamma)
```

Other than that, the changes look good to me. :) 

I'll leave it to @prabhat00155 to do the final checks on our side and merge when ready.

**Review by prabhat00155 — APPROVED** [review]
Thanks @xiaohu2015!

**Pull request merged**